File: JS/Tree/110.Balanced Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
const maxLength = function(node) {
    if(node) {
        return 1 + Math.max(maxLength(node.left), maxLength(node.right));
    }
    return 0;
}
var isBalanced = function(root) {
    if (!root || (!root.left && !root.right)) {
        return true; 
    }
    return Math.abs(maxLength(root.left) - maxLength(root.right)) < 2 && isBalanced(root.left) && isBalanced(root.right);
};

File: Golang/Tree/110.balanced-binary-tree.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isBalanced(root *TreeNode) bool {
	var getDepth func(node *TreeNode) int
	var checkBalanced func(node *TreeNode) bool
	getDepth = func(node *TreeNode) int {
		if node != nil {
			left := getDepth(node.Left)
			right := getDepth(node.Right)
			if left > right {
				return 1 + left
			}
			return 1 + right
		}
		return 0
	}
	checkBalanced = func(node *TreeNode) bool {
		if node != nil {
			left := getDepth(node.Left)
			right := getDepth(node.Right)
			if int(math.Abs(float64(left-right))) < 2 {
				return checkBalanced(node.Left) && checkBalanced(node.Right)
			}
			return false
		}
		return true
	}
	return checkBalanced(root)
}

func isBalancedOptimized(root *TreeNode) bool {
	var getDepth func(node *TreeNode) int
	getDepth = func(node *TreeNode) int {
		if node == nil {
			return 0
		}
		left := getDepth(node.Left)
		right := getDepth(node.Right)
		if left == -1 || right == -1 {
			return -1
		}
		if int(math.Abs(float64(left-right))) > 1 {
			return -1
		}
		if left > right {
			return 1 + left
		}
		return 1 + right
	}
	return getDepth(root) != -1
}
